src/firestore/indexes-sort.ts

```diff
@@ -0,0 +1,173 @@
+export function compareApiIndex(a: API.Index, b: API.Index): number {
+  // When these indexes are used as part of a field override, the name is
+  // not always present or relevant.
+  if (a.name && b.name) {
+    const aName = util.parseIndexName(a.name);
+    const bName = util.parseIndexName(b.name);
+
+    if (aName.collectionGroupId !== bName.collectionGroupId) {
+      return aName.collectionGroupId.localeCompare(bName.collectionGroupId);
+    }
+  }
+
+  if (a.queryScope !== b.queryScope) {
+    return compareQueryScope(a.queryScope, b.queryScope);
+  }
+
+  return compareArrays(a.fields, b.fields, compareIndexField);
+}
+
+/**
+ * Compare two Field api entries for sorting.
+ *
+ * Comparisons:
+ *   1) The collection group.
+ *   2) The field path.
+ *   3) The indexes list in the config.
+ */
+export function compareApiField(a: API.Field, b: API.Field): number {
+  const aName = util.parseFieldName(a.name);
+  const bName = util.parseFieldName(b.name);
+
+  if (aName.collectionGroupId !== bName.collectionGroupId) {
+    return aName.collectionGroupId.localeCompare(bName.collectionGroupId);
+  }
+
+  if (aName.fieldPath !== bName.fieldPath) {
+    return aName.fieldPath.localeCompare(bName.fieldPath);
+  }
+
+  return compareArrays(a.indexConfig.indexes || [], b.indexConfig.indexes || [], compareApiIndex);
+}
+
+/**
+ * Compare two Field override specs for sorting.
+ *
+ * Comparisons:
+ *   1) The collection group.
+ *   2) The field path.
+ *   3) The list of indexes.
+ */
+export function compareFieldOverride(a: Spec.FieldOverride, b: Spec.FieldOverride): number {
+  if (a.collectionGroup !== b.collectionGroup) {
+    return a.collectionGroup.localeCompare(b.collectionGroup);
+  }
+
+  if (a.fieldPath !== b.fieldPath) {
+    return a.fieldPath.localeCompare(b.fieldPath);
+  }
+
+  return compareArrays(a.indexes, b.indexes, compareFieldIndex);
+}
+
+/**
+ * Compare two IndexField objects.
+ *
+ * Comparisons:
+ *   1) Field path.
+ *   2) Sort order (if it exists).
+ *   3) Array config (if it exists).
+ */
+function compareIndexField(a: API.IndexField, b: API.IndexField): number {
+  if (a.fieldPath !== b.fieldPath) {
+    return a.fieldPath.localeCompare(b.fieldPath);
+  }
+
+  if (a.order !== b.order) {
+    return compareOrder(a.order, b.order);
+  }
+
+  if (a.arrayConfig !== b.arrayConfig) {
+    return compareArrayConfig(a.arrayConfig, b.arrayConfig);
+  }
+
+  return 0;
+}
+
+function compareFieldIndex(a: Spec.FieldIndex, b: Spec.FieldIndex): number {
+  if (a.queryScope !== b.queryScope) {
+    return compareQueryScope(a.queryScope, b.queryScope);
+  }
+
+  if (a.order !== b.order) {
+    return compareOrder(a.order, b.order);
+  }
+
+  if (a.arrayConfig !== b.arrayConfig) {
+    return compareArrayConfig(a.arrayConfig, b.arrayConfig);
+  }
+
+  return 0;
+}
+
+function compareQueryScope(a: API.QueryScope, b: API.QueryScope): number {
+  const sequence = [API.QueryScope.COLLECTION_GROUP, API.QueryScope.COLLECTION, undefined];
```

Comment:
these are being created every time the function is called. maybe they should be pulled out into a constant a the top level? (same below x2)